samples/bluetooth/hci_vs_scan_req/src/main.c

```diff
@@ -0,0 +1,326 @@
+};
+
+#elif defined(CONFIG_BT_CTLR_VS_SCAN_REQ_RX)
+
+static bool vs_scanned(struct net_buf_simple *buf)
+{
+	struct bt_hci_evt_vs_scan_req_rx *evt;
+	struct bt_hci_evt_vs *vs;
+
+	vs = net_buf_simple_pull_mem(buf, sizeof(*vs));
+	evt = (void *)buf->data;
+
+	printk("%s subevent 0x%02x peer %s rssi %d\n", __func__,
+	       vs->subevent, bt_addr_le_str(&evt->addr), evt->rssi);
+
+	return true;
+}
+
+#endif /* defined(CONFIG_BT_EXT_ADV) */
+
+#if defined(CONFIG_BT_EXT_ADV)
+static void create_advertising(void)
+{
+	int err;
+
+	/* Create a scannable advertising set */
+	err = bt_le_ext_adv_create(&parameters, &ext_adv_callbacks, &ext_adv);
+	if (err) {
+		printk("Create advertising set err %d\n", err);
+		return;
+	}
+
+	/* Set's advertising or scan response data */
+	err = bt_le_ext_adv_set_data(ext_adv, advertisement_data, ARRAY_SIZE(advertisement_data),
+				     scan_response_data, ARRAY_SIZE(scan_response_data));
+	if (err) {
+		printk("Set adv or scan response data err %d\n", err);
+		return;
+	}
+
+	printk("Advertising set successfully created (%s)\n", CONFIG_BT_DEVICE_NAME);
+}
+#endif /* defined(CONFIG_BT_EXT_ADV) */
+
+static int start_advertising(void)
+{
+	int err;
+
+#if defined(CONFIG_BT_CTLR_VS_SCAN_REQ_RX)
+	err = bt_hci_register_vnd_evt_cb(vs_scanned);
+	enable_legacy_adv_scan_request_event(true);
+	err = bt_le_adv_start(&parameters, advertisement_data, ARRAY_SIZE(advertisement_data),
+			      scan_response_data, ARRAY_SIZE(scan_response_data));
+	if (err) {
+		printk("Start legacy adv err %d\n", err);
+		return err;
+	}
+#elif defined(CONFIG_BT_EXT_ADV)
+	err = bt_le_ext_adv_start(ext_adv, BT_LE_EXT_ADV_START_DEFAULT);
+	if (err) {
+		printk("Start ext adv err %d\n", err);
+		return err;
+	}
+#endif /* defined(CONFIG_BT_CTLR_VS_SCAN_REQ_RX) */
+	printk("Advertising successfully started (%s)\n", CONFIG_BT_DEVICE_NAME);
+	return 0;
+}
+
+static void periodic_checks(void)
+{
+#if defined(CONFIG_BT_PERIPHERAL) && defined(CONFIG_BT_EXT_ADV)
+	if (disconnect_event_happened) {
+		int rv = start_advertising();
+
+		if (rv != 0) {
+			printk("start adv err: %d\n", rv);
+		} else {
+			disconnect_event_happened = false;
+		}
+	}
+#endif /* defined(CONFIG_BT_PERIPHERAL) && defined(CONFIG_BT_EXT_ADV) */
+}
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	if (err) {
+		printk("Connection failed (err 0x%02x)\n", err);
+	} else {
+		bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+		printk("Connected at %s\n", addr);
+	}
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	printk("Disconnected (reason 0x%02x)\n", reason);
+
+#if defined(CONFIG_BT_PERIPHERAL)
+#if defined(CONFIG_BT_EXT_ADV)
+	disconnect_event_happened = true;
+#else
+	(void) start_advertising(); /* gives errno -12 = -ENOMEM with if CONFIG_BT_EXT_ADV */
+#endif
+#endif
+}
+
+BT_CONN_CB_DEFINE(conn_callbacks) = {
+	.connected = connected,
+	.disconnected = disconnected,
+};
```

Comment:
Indeed in our application the equivalent module does exactly as you say. I lost the guard when porting part of that code in the sample app, even because the order is different due to different coding rules. Fixed now.